File: src/main/java/uk/edu/populationfitness/models/genes/bitset/CachingInterpolatingBitSetGenes.java
package uk.edu.populationfitness.models.genes.bitset;

import uk.edu.populationfitness.models.Config;

/**
 * Only calculates the fitness if the fitness factor has changed.
 */
public abstract class CachingInterpolatingBitSetGenes extends InterpolatingBitSetGenes{

    CachingInterpolatingBitSetGenes(Config config, double maxInterpolatedValue) {
        super(config, maxInterpolatedValue);
    }

    /**
     * Implement this to calculate the fitness from an array of double unknowns.
     *
     * @param unknowns
     * @return
     */
    protected abstract double calculateFitnessFromGenes(double[] unknowns);

    @Override
    public double fitness() {
        if (isFitnessStored()) {
            return storedFitness();
        }

        return storeInvertedFitness(calculateFitnessFromGenes(asDoubles()));
    }
}

File: DotNet/PopulationFitness/PopulationFitness/Models/Genes/BitSet/CachingInterpolatingBitSetGenes.cs
namespace PopulationFitness.Models.Genes.BitSet
{
    /**
     * Only calculates the fitness if the fitness factor has changed.
     */
    public abstract class CachingInterpolatingBitSetGenes : InterpolatingBitSetGenes
    {

        protected CachingInterpolatingBitSetGenes(Config config, double maxInterpolatedValue) : base(config, maxInterpolatedValue)
        {
        }

        /**
         * Implement this to calculate the fitness from an array of integers.
         *
         * @param integer_values
         * @return
         */
        protected abstract double CalculateFitnessFromIntegers(long[] integer_values);

        public override double Fitness
        {
            get
            {
                if (IsFitnessStored())
                {
                    return StoredFitness();
                }

                return StoreInvertedFitness(CalculateFitnessFromIntegers(AsIntegers));
            }
        }
    }
}
